pipeline/transformer/transformer_calm/src/main/scala/uk/ac/wellcome/platform/transformer/calm/Worker.scala

```diff
@@ -1,38 +1,34 @@
 package uk.ac.wellcome.platform.transformer.calm
 
-import uk.ac.wellcome.messaging.sns.NotificationMessage
-import uk.ac.wellcome.messaging.sqs.SQSStream
-import uk.ac.wellcome.json.JsonUtil._
-import uk.ac.wellcome.platform.transformer.calm.service.{
-  CalmTransformerWorkerServiceState,
-  MessageReadError
-}
-import uk.ac.wellcome.storage.Version
-
-trait Worker[Store, StoreKey, SourceData, SinkData] extends Runnable {
-  type Result[T] = Either[Throwable, T]
+import akka.{Done, NotUsed}
+import akka.stream.scaladsl.{Flow, Sink, Source}
 
-  def runStream(stream: SQSStream[NotificationMessage]) =
-    stream.foreach(this.getClass.getSimpleName, processMessage)
+import scala.concurrent.Future
 
-  def processMessage(message: NotificationMessage) =
-    for {
-      key <- readMessage(message)
-      storeRecord <- get(key)
-      sinkData <- doWork(storeRecord)
-      sent <- done(sinkData)
-    } yield sent
+trait Worker[Message, MessageData, SinkData] {
+  private type Result[T] = Either[Throwable, T]
+  def errorSink: Sink[Result[_], Future[Done]] = Sink.ignore
 
-  def get(storeKey: StoreKey): Result[SourceData]
+  def processMessage(source: Source[Message, NotUsed]) =
+    source.via(decodeMessageFlow).via(workFlow).via(doneFlow)
 
-  private def readMessage(message: NotificationMessage): Result[StoreKey] =
-    fromJson[StoreKey](message.body).toEither match {
-      case Left(err) =>
-        Left(new Exception(s"Worker: Cannot read message, ${err}"))
-      case _ => _
+  private def divertEither[I, O](
+    flow: Flow[I, Result[O], NotUsed],
+    to: Sink[Result[_], Future[Done]]): Flow[I, O, NotUsed] =
+    flow.divertTo(Sink.ignore, _.isLeft).collect {
+      case Right(out) => out
     }
 
-  def doWork(sourceData: SourceData): Result[SinkData]
+  val decodeMessageFlow: Flow[Message, MessageData, NotUsed] =
+    divertEither(Flow.fromFunction(decodeMessage), to = errorSink)
+
+  val workFlow: Flow[MessageData, SinkData, NotUsed] =
+    divertEither(Flow.fromFunction(work), to = errorSink)
+
+  val doneFlow: Flow[SinkData, Unit, NotUsed] =
+    divertEither(Flow.fromFunction(done), to = errorSink)
 
+  def decodeMessage(message: Message): Result[MessageData]
+  def work(messageData: MessageData): Result[SinkData]
   def done(sinkData: SinkData): Result[Unit]
```

Comment:
Any reason we don't allow the implementer to specify a `Flow` for these here, instead of just a function? That would allow easier concurrency / async stuff etc, and provide more fine grained backpressure rather than for the stage as a whole

Reply:
Could do when it becomes a thing I suppose.

The reason I haven't is it feels like I have abstracted as far back as might be useful over as far back as I can.

Another reason is to move the Flow stuff out. You just need to give us the business logic. 

Maybe having a `WorkerWork` and a `WorkerFlow` that you can compose might be useful. I am thinkinng of how to split this up into more composable parts, and maybe this is another one.

Reply:
But then I don't really see the advantage of using akka-streams at all? As it is currently it is just a sequential functions chained together, the streams stuff mainly gets useful when there is async stuff or heavier computations etc, and this interface doesn't give any control over how any of that is executed

Reply:
It's not a bad point, we could just wrap them in the [for comprehension as before](https://github.com/wellcomecollection/catalogue/pull/399/files#diff-143261f81c47872cc717d20e7a22c886R47-R52).

The point of this abstraction was to be able have an interface defined on a Worker that read a message, did some work, and then did something with the output of that work, and keeping any tracing of that interface outside of the implementation.

I was going to use the graph like interface for ☝️ in terms of tracing.

It feels like each step would need to have it's own configuration, at which point you're basically re-writing the trait everytime you want to use it.

Saying that, this is a good point, and I'll have a little stab at how it might look.

Reply:
Fwiw I think we should use akka streams, as it is really nice way to express the sort of data flow computations we work with (especially when it does get more complicated with calls to other services etc)..

Maybe we could define the "higher level" function interface like we have here, but also a "lower level" stream interface which the other is based on (this would be the one used by e.g. [`CalmAdapterWorkerService`](https://github.com/wellcomecollection/catalogue/blob/master/calm_adapter/calm_adapter/src/main/scala/uk/ac/wellcome/calm_adapter/CalmAdapterWorkerService.scala), [`MetsAdapterWorkerService`](https://github.com/wellcomecollection/catalogue/blob/master/mets_adapter/mets_adapter/src/main/scala/uk/ac/wellcome/mets_adapter/services/MetsAdapterWorkerService.scala), which do work on streams)

Reply:
Another though: I think lots of our current worker services would have to use an interface which at least allows returning of `Future[_]`, as they do async things. Otherwise we are going to have to `Await` everywhere, which I don't think we want

Reply:
Did this as a first pass?

Reply:
Was thinking we could also:

```scala
/** This worker:
  * - Reads from a VHS, and decodes to `VHSDataType`
  * - Transforms it into a `TransformedBaseWork`
  * - Emits the work via `BigMessageSender`
  * */
class VhsReaderWorker[VHSDataType]
    extends StreamWorker[
      (Message, NotificationMessage),
      VHSDataType,
      TransformedBaseWork] {

  type Result[T] = Either[Throwable, T]
  type StoreKey = Version[String, Int]
}
```

But we land up coupling in the name of abstraction...